### PR TITLE
[Notion] Implement an ability to retry for failed paginated api call

### PR DESCRIPTION
**Pull request by praveen-elastic**
## Part of #1958 

During performance testing, it was reported that the [notion sdk module](https://github.com/ramnes/notion-sdk-py/blob/d5e5f5c98ae5578ce4b0130b65ad6e8f9e92e02b/notion_client/helpers.py#L58) does not provision any retry mechanism for a paginated api call made via `async_iterate_paginated_api`. 

This PR attempts to implement `async_iterate_paginated_api` with an ability to retry a failed page. 

Simply decorating the module's helper method [async_iterate_paginated_api](https://github.com/ramnes/notion-sdk-py/blob/d5e5f5c98ae5578ce4b0130b65ad6e8f9e92e02b/notion_client/helpers.py#L58) with retryable decorator won't work as that would mean any error thrown by page k (k >1) would result in retrying from page 1 instead of retry on page k.

#### Pre-Review Checklist
- [x] this PR has a meaningful title
- [x] this PR links to all relevant github issues that it fixes or partially addresses
- [x] if there is no GH issue, please create it. Each PR should have a link to an issue
- [x] this PR has a thorough description
- [x] Covered the changes with automated tests
- [x] Tested the changes locally
- [x] Added a label for each target release version (example: `v7.13.2`, `v7.14.0`, `v8.0.0`)
- [ ] Considered corresponding documentation changes
- [ ] Contributed any configuration settings changes to the configuration reference
- [ ] if you added or changed Rich Configurable Fields for a Native Connector, you made a corresponding PR in [Kibana](https://github.com/elastic/kibana/blob/main/packages/kbn-search-connectors/types/native_connectors.ts)

#### Changes Requiring Extra Attention

- [ ] Security-related changes (encryption, TLS, SSRF, etc)
- [ ] New external service dependencies added.

### CODE REVIEW

File: connectors/sources/notion.py [hunk 1]
```diff
@@ -96,6 +96,48 @@ async def get_via_session(self, url):
...
+        self, function: Callable[..., Awaitable[Any]], next_cursor=None, **kwargs: Any
+    ):
+        try:
+            return await function(start_cursor=next_cursor, **kwargs)
+        except Exception as exception:
+            if exception.code == "rate_limited" and exception.status == 429:
+                retry_after = (
+                    exception.headers.get("retry-after") or DEFAULT_RETRY_SECONDS
+                )
+                self._logger.info(
```
Comment (seanstory) [inline connectors/sources/notion.py:115]: let's include in the logs what the request was. I assume that's captured in the `next_cursor` or the `kwargs`, but capturing the `function`'s name might be good too.

File: connectors/sources/notion.py [hunk 2]
```diff
@@ -96,6 +96,48 @@ async def get_via_session(self, url):
...
+        strategy=RetryStrategy.EXPONENTIAL_BACKOFF,
+        skipped_exceptions=NotFound,
+    )
+    async def fetch_results(
+        self, function: Callable[..., Awaitable[Any]], next_cursor=None, **kwargs: Any
+    ):
+        try:
+            return await function(start_cursor=next_cursor, **kwargs)
+        except Exception as exception:
+            if exception.code == "rate_limited" and exception.status == 429:
```
Comment (seanstory) [inline connectors/sources/notion.py:111]:
```suggestion
            if exception.code == "rate_limited" or exception.status == 429:
```

File: connectors/sources/notion.py [hunk 3]
```diff
@@ -96,6 +96,48 @@ async def get_via_session(self, url):
...
                 raise
 
+    @retryable(
+        retries=RETRIES,
+        interval=RETRY_INTERVAL,
+        strategy=RetryStrategy.EXPONENTIAL_BACKOFF,
+        skipped_exceptions=NotFound,
+    )
+    async def fetch_results(
+        self, function: Callable[..., Awaitable[Any]], next_cursor=None, **kwargs: Any
```
Comment (seanstory) [inline connectors/sources/notion.py:106]: why is `next_cursor` an explicit arg, instead of being included in `kwargs`? Below, it looks like this starts as 
```
kwargs.pop("start_cursor", None)
```
so I don't see the point.

File: connectors/sources/notion.py [hunk 4]
```diff
@@ -96,6 +96,48 @@ async def get_via_session(self, url):
...
+            try:
+                response = await self.fetch_results(function, next_cursor, **kwargs)
+                if response:
+                    for result in response.get("results"):
+                        yield result
+                    next_cursor = response.get("next_cursor")
+                    if not response["has_more"] or next_cursor is None:
+                        return
+            except Exception as error:
+                raise error
```
Comment (seanstory) [inline connectors/sources/notion.py:139]: This seems entirely redundant. Unless you're going to handle `error` or log something, why bother catching it just to re-raise?

File: tests/sources/test_notion.py [hunk 5]
```diff
@@ -740,13 +759,37 @@ async def test_advanced_rules_validation(advanced_rules, expected_validation_res
...
         ):
             validation_result = await NotionAdvancedRulesValidator(source).validate(
                 advanced_rules
             )
-        assert validation_result == expected_validation_result
+            assert validation_result == expected_validation_result
+
+
+@pytest.mark.asyncio
+async def test_async_iterate_paginated_api():
```
Comment (seanstory) [inline tests/sources/test_notion.py:776]: Can you add tests to ensure that 429's are retried, but other errors are not? Perhaps also add a test to make sure that the original `async_iterate_paginated_api` is not being called anywhere?

File: connectors/sources/notion.py [hunk 6]
```diff
@@ -28,7 +28,7 @@
 RETRY_INTERVAL = 2
 DEFAULT_RETRY_SECONDS = 30
 BASE_URL = "https://api.notion.com"
-MAX_CONCURRENT_CLIENT_SUPPORT = 3
+MAX_CONCURRENT_CLIENT_SUPPORT = 30
```
Comment (seanstory) [inline connectors/sources/notion.py:31]: This seems like a large change and looks unrelated. What's the rationale?

File: connectors/sources/notion.py [hunk 7]
```diff
@@ -423,7 +465,7 @@ def get_default_configuration(cls):
                 "value": False,
             },
             "concurrent_downloads": {
-                "default_value": 20,
+                "default_value": 30,
```
Comment (seanstory) [inline connectors/sources/notion.py:468]: Why's this being changed?

File: connectors/sources/notion.py (hunk 3 above)
Comment (praveen-elastic) [reply connectors/sources/notion.py:106]: the `next_cursor` passed to `fetch_results` method is only equal to `start_cursor` during 1st iteration, hence that can be fetched via `kwargs.pop("start_cursor", None)`. 

But in the subsequent iterations of while loop, the `next_cursor` is available in the response of previous call. Hence, we cannot rely on `kwargs` for these iterations. Therefore we need to pass that explicitly along with `kwargs`

File: tests/sources/test_notion.py (hunk 5 above)
Comment (praveen-elastic) [reply tests/sources/test_notion.py:776]: Updated via commit id: [2a601f1](https://github.com/elastic/connectors/pull/2297/commits/2a601f1935fd829667106e11af1621eac2c54482)

File: connectors/sources/notion.py (hunk 6 above)
Comment (praveen-elastic) [reply connectors/sources/notion.py:31]: The downloading of the attachments happens via `aiohttp` session using public urls which expire in 1 hour, also those aren't restricted by Notion's rate limit of 3, so at a time more connections can be done for downloading files.

Hence following changes were made for `MAX_CONCURRENT_CLIENT_SUPPORT` and `concurrent_downloads` that is specifically used for downloading of files only.

Should I raise a separate PR for this ?

File: connectors/sources/notion.py (hunk 7 above)
Comment (praveen-elastic) [reply connectors/sources/notion.py:466]: same as the reason above, it was changed for file downloads.

File: connectors/sources/notion.py (hunk 6 above)
Comment (seanstory) [reply connectors/sources/notion.py:31]: No, I'm fine with you including small fixes in larger PRs. In the future, just try to call them out either with code comments or in the PR description, so it's clear what's being changed and why. 👍

File: tests/sources/test_notion.py [hunk 8]
```diff
@@ -740,13 +759,85 @@ async def test_advanced_rules_validation(advanced_rules, expected_validation_res
...
+        assert exc_info.value.code == "object_not_found"
+        assert exc_info.value.status == 404
+
+
+@pytest.mark.asyncio
+async def test_original_async_iterate_paginated_api_not_called():
+    with patch(
+        "notion_client.helpers.async_iterate_paginated_api"
+    ) as mock_async_iterate_paginated_api:
+        assert not mock_async_iterate_paginated_api.called
```
Comment (seanstory) [inline tests/sources/test_notion.py:843]: This isn't testing anything. You'd need to have the sync actually run inside the `with patch`.

File: tests/sources/test_notion.py [hunk 9]
```diff
@@ -740,13 +759,85 @@ async def test_advanced_rules_validation(advanced_rules, expected_validation_res
...
+            message="Rate limit exceeded.",
+            response=Response(status_code=429, text="Rate limit exceeded."),
+        )
+
+    async with create_source(
+        NotionDataSource, notion_secret_key="secert_key"
+    ) as source:
+        with patch("connectors.sources.notion.DEFAULT_RETRY_SECONDS", 0.3):
+            with pytest.raises(Exception) as exc_info:
+                await source.notion_client.fetch_results(mock_function_with_429)
```
Comment (seanstory) [inline tests/sources/test_notion.py:813]: Can you adapt this to show that it raises if it keeps getting the 429, but explicitly _does not_ raise an error if it 429s once, but then responds successfully after the retry? For example see https://github.com/elastic/connectors/blob/main/tests/sources/test_sharepoint_online.py#L571

File: tests/sources/test_notion.py [hunk 10]
```diff
@@ -793,3 +793,51 @@ async def mock_function(**kwargs):
...
+
+    async with create_source(
+        NotionDataSource, notion_secret_key="secert_key"
+    ) as source:
+        with pytest.raises(APIResponseError) as exc_info:
+            await source.notion_client.fetch_results(mock_function_with_other_error)
+
+        assert exc_info.value.code == "object_not_found"
+        assert exc_info.value.status == 404
+
```
Comment (seanstory) [inline tests/sources/test_notion.py:836]: I don't think this is currently asserting a lack of retries. Can you assert how many times the request is actually made to show that a retry is not occurring?

**Review by seanstory — APPROVED** [review]
